File: services/ctl-api/internal/app/runbooks/service/get_install_runbooks.go
package service

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"github.com/nuonco/nuon/services/ctl-api/internal/app"
	"github.com/nuonco/nuon/services/ctl-api/internal/pkg/cctx"
	"github.com/nuonco/nuon/services/ctl-api/internal/pkg/db"
	"github.com/nuonco/nuon/services/ctl-api/internal/pkg/db/scopes"
)

// @ID				GetInstallRunbooks
// @Summary		get runbooks for an install
// @Tags			runbooks
// @Accept			json
// @Produce		json
// @Security		APIKey
// @Security		OrgID
// @Param			install_id	path	string	true	"install ID"
// @Param			offset		query	int		false	"offset"	Default(0)
// @Param			limit		query	int		false	"limit"		Default(10)
// @Success		200			{array}	app.InstallRunbook
// @Failure		400			{object}	stderr.ErrResponse
// @Failure		401			{object}	stderr.ErrResponse
// @Failure		403			{object}	stderr.ErrResponse
// @Failure		404			{object}	stderr.ErrResponse
// @Failure		500			{object}	stderr.ErrResponse
// @Router			/v1/installs/{install_id}/runbooks [get]
func (s *service) GetInstallRunbooks(ctx *gin.Context) {
	enabled, err := s.featuresClient.FeatureEnabled(ctx, app.OrgFeatureRunbooks)
	if err != nil || !enabled {
		ctx.Error(fmt.Errorf("runbooks feature is not enabled"))
		return
	}

	installID := ctx.Param("install_id")
	org, err := cctx.OrgFromContext(ctx)
	if err != nil {
		ctx.Error(err)
		return
	}

	installRunbooks := []*app.InstallRunbook{}
	res := s.db.WithContext(ctx).
		Scopes(scopes.WithOffsetPagination).
		Joins("JOIN runbooks ON runbooks.id = install_runbooks.runbook_id AND runbooks.deleted_at = 0").
		Preload("Runbook").
		Preload("Runbook.Configs", func(tx *gorm.DB) *gorm.DB {
			return tx.Scopes(scopes.WithOverrideTable("runbook_configs_latest_view_v1"))
		}).
		Preload("Runbook.Configs.Steps", func(tx *gorm.DB) *gorm.DB {
			return tx.Order("idx ASC")
		}).
		Preload("Runs", func(tx *gorm.DB) *gorm.DB {
			return tx.Scopes(scopes.WithOverrideTable("install_runbook_runs_latest_view_v1"))
		}).
		Where(app.InstallRunbook{OrgID: org.ID, InstallID: installID}).
		Order("install_runbooks.created_at DESC").
		Find(&installRunbooks)
	if res.Error != nil {
		ctx.Error(fmt.Errorf("unable to get install runbooks: %w", res.Error))
		return
	}

	installRunbooks, err = db.HandlePaginatedResponse(ctx, installRunbooks)
	if err != nil {
		ctx.Error(fmt.Errorf("unable to handle paginated response: %w", err))
		return
	}

	ctx.JSON(http.StatusOK, installRunbooks)
}

File: services/dashboard-ui/client/lib/ctl-api/installs/runbooks/get-install-runbooks.ts
import { api } from '@/lib/api'
import type { TPaginationParams } from '@/types'
import { buildQueryParams } from '@/utils/build-query-params'
import type { TRunbook } from '@/lib/ctl-api/apps/runbooks'

export type TInstallRunbook = {
  id: string
  install_id?: string
  runbook_id?: string
  org_id?: string
  created_at?: string
  updated_at?: string
  status?: string
  runbook: TRunbook
  runs?: TInstallRunbookRun[]
}

export type TInstallRunbookRun = {
  id: string
  install_id?: string
  install_runbook_id?: string
  runbook_config_id?: string
  org_id?: string
  created_at?: string
  updated_at?: string
  status?: string
  status_description?: string
  triggered_by_id?: string
  created_by_id?: string
  created_by?: {
    email?: string
    name?: string
  }
  install_workflow_id?: string | null
  install_workflow?: {
    id: string
    status?: {
      status?: string
      status_human_description?: string
    }
  } | null
}

export async function getInstallRunbooks({
  installId,
  orgId,
  limit,
  offset,
  q,
}: {
  installId: string
  orgId: string
  q?: string
} & TPaginationParams) {
  return api<TInstallRunbook[]>({
    orgId,
    path: `installs/${installId}/runbooks${buildQueryParams({ limit, offset, q })}`,
    paginated: true,
  })
}
